### PR TITLE
fix ConfirmSubscription access to SNS store

### PR DESCRIPTION
**Pull request by bentsku**
As reported in #8118, there was an issue with confirming a subscription from a simple http client, because we would use the region from the context, which was using the default region of LocalStack (set from the auth handler `MissingAuthHeaderInjector`). However, we should use the region from the `TopicARN`. I've re-validated the tests against AWS with a local setup with https://localhost.run, and could validate different behaviours that were untested. This adds more parity into our HTTP subscriptions, and fix the linked issue. 

Also, I've sneaked a little store change to get the parity needed. This should not break persistence nor cloud pods as tokens are usually short lived, and it is more additive. We won't try to get a topic ARN anymore from this dict, so it would not create error per se. 

_fixes #8118_

### CODE REVIEW

File: localstack/services/sns/provider.py [hunk 1]
```diff
@@ -404,14 +404,33 @@ def confirm_subscription(
...
+        if parsed_arn["region"] != get_region_from_subscription_token(token):
+            raise InvalidParameterException("Invalid parameter: Topic")
 
-        subscription_arn = topic_tokens.pop(token, None)
+        subscription_arn = store.subscription_tokens.get(token)
         if not subscription_arn:
             raise InvalidParameterException("Invalid parameter: Token")
 
         subscription = store.subscriptions.get(subscription_arn)
+        if not subscription:
```
Comment (bentsku) [inline localstack/services/sns/provider.py:426]: I should add a todo here. It might be that confirming a deleted subscription might actually recreate it, from the description of `UnsubscribeURL` messages. However, this is out of scope of the PR.

File: tests/unit/test_sns.py [hunk 2]
```diff
@@ -579,3 +583,19 @@ def test_filter_policy_on_message_body(self):
...
+        token = encode_subscription_token_with_region(region)
+        assert len(token) == 64
+        token_region = get_region_from_subscription_token(token)
+        assert token_region == region
+
+    @pytest.mark.parametrize(
+        "token", ["abcdef123", "mynothexstring", "us-west-2", b"test", b"test2f", "test2f"]
+    )
+    def test_decode_token_with_no_region_encoded(self, token):
+        with pytest.raises(Exception) as e:
```
Comment (dave-mclean) [inline tests/unit/test_sns.py:598]:
```suggestion
        with pytest.raises(InvalidParameterException) as e:
```
I think being more specific about what exception is raised here might catch possible errors in future.

Comment (bentsku) [reply tests/unit/test_sns.py:598]: I am matching on the exception under, and if the exception raised does not have the following message `"Invalid parameter: Token"`, the test will fail anyway. Agreed that it could be safer though, I'll add it after the second review, thanks for pointing this out!

**Review by thrau — COMMENTED** [review]
Thanks for tackling this so quickly @bentsku. Amazing test coverage as always!

Just one comment/question I think needs clarification:
> This should not break persistence nor cloud pods as tokens are usually short lived, and it is more additive.

i couldn't find code that deletes the `subscription_tokens`. there was one instance of `subscription_tokens.pop` that was removed - so it seems to me that, as soon as you have one subscription in your SNS pod, the `subscription_tokens` will be populated with state, that will be serialized into the store.
persistence will definitely break in that case, since we have `Dict[str, Dict[str, str]]` in the attribute, but now expect `Dict[str, str]`

if i'm missing something here and am wrong then let's go ahead and merge for the patch release, otherwise i think it would be good to stage it for 2.1

**Comment by bentsku** [comment]
Yes, the persistence was a bit of a pickle.
Before, we would delete the token right away, but it wasn't the right behaviour as we still need to be able to match a token to a subscription.
So in my head, in existent cloud pods / persistence, there should not be any token saved (famous last words), or it would be at an unfortunate moment, between creating a subscription and confirming it. Only newly created pods will always have subscription tokens. 

I guess it boils down to a question I didn't have the answer: when restoring persistence, will the new store keep the old shape, or be cast to the new shape? And does the shape really have an incidence?
Because let's say the old store looks like this:
```python
subscription_tokens = {
  "faketopicarn": {
    "token": "subscriptionarn"
  }
}
# this would be old shape that is still in the store after restore
```
When accessing that store, we would just try to get the token from it and it would not exist basically, because it's one level deeper. We would break this exact moment of creating a subscription, saving the pod, restoring the pod, validating the subscription. I guess that's a very specific scenario. 

But if that's still a risk, we can stage it for 2.1, no problem for me! Might be safer, to avoid any issue.

File: tests/unit/test_sns.py (hunk 2 above)
Comment (dave-mclean) [reply tests/unit/test_sns.py:598]: ah interesting, I didn't see the assert underneath - sorry!

I imagine you already know this (and therefore have a reason for not doing it!) but just in case you can combine that `match` with `pytest.raises` into one line like this:

```
with pytest.raises(InvalidParameterException, match="Invalid parameter: Token"):
    get_region_from_subscription_token(token)
```

Comment (bentsku) [reply tests/unit/test_sns.py:598]: Thanks for tip! I guess I like having explicit `assert` throughout the test, to quickly spot what we are looking for. But that's a good tip, thanks!
